### ISSUE TITLE
Defined variables become unbound in unreachable code

### ISSUE THREAD

**Issue by has2k1** [microsoft/pyright#7189]
**Describe the bug**

pyright sees as unbound the variables that are defined and then later referred to in theoretically unreachable code. Variables that get into scope as parameters are not affected.

**Code or Screenshots**

```python
def func1(param: int | float):
    err_msg = f"{param=} is a bad value"
    if isinstance(param, int):
        return 1
    elif isinstance(param, float):
        return 1.0
    raise ValueError(err_msg)  # "err_msg" is unbound [reportUnboundVariable]


def func2(param: int | float):
    err_msg = f"{param=} is a bad value"
    if isinstance(param, int):
        raise ValueError("int")
    elif isinstance(param, float):
        raise ValueError("float")
    raise ValueError(err_msg)  #  "err_msg" is unbound [reportUnboundVariable]


def func3(param: int | float, err_msg: str):
    if isinstance(param, int):
        return 1
    elif isinstance(param, float):
        return 1.0
    raise ValueError(err_msg)
```

At the end of `func1` and `func2`, `err_msg` is seen as unbound. `func3` has no problem(correctly).

**Comment by erictraut** [comment]
This is working as designed, but I can understand why it would be confusing. I'll think about whether there's a better approach here.

The problem is that the variable has no type annotation, so its type must be determined fully by code flow analysis — based on prior assignments in the code flow graph. However, there are no reachable assignments when walking the graph backward from the final line of code. No reachable assignments effectively means that the variable has no bound value.

**Comment by erictraut** [comment]
I found a good way to handle this. Not only does it eliminate the error, but it also results in the last statements being displayed as "dimmed" so it's clear that they are unreachable. This will be included in the next release.

**Comment by has2k1** [comment]
> Not only does it eliminate the error, but it also results in the last statements being displayed as "dimmed" so it's clear that they are unreachable.

Great. Though, this is interesting:

> The problem is that the variable has no type annotation, so its type must be determined fully by code flow analysis — based on prior assignments in the code flow graph. However, there are no reachable assignments when walking the graph backward from the final line of code.

My mental modal was, type information always flowing downwards and sifted (set difference) or gathered up (set unions). But this means for annotated variables the inference is top-down (with set differences) set, but for unannotated variables it is bottom-up (with set unions).

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.350, which I just published.
